### PR TITLE
symmetric eigensolver

### PR DESCRIPTION
**Pull request by djelovina**
depends on householder tridiagonalisation and tridiagonal eigensolver complex-hermitian version will use the same call

### CODE REVIEW

File: tests/smoke/alp_zheevd.cpp [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+#endif
+
+#include <alp.hpp>
+#include <alp/algorithms/householder_tridiag.hpp>
+#include <alp/algorithms/symm_tridiag_eigensolver.hpp>
+#include <graphblas/utils/iscomplex.hpp> // use from grb
+#include "../utils/print_alp_containers.hpp"
+
+//once TEMPDISABLE is removed the code should be in the final version
+#define TEMPDISABLE
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:33]: Is this still used?

File: tests/smoke/alp_zheevd.cpp [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+#endif
+
+constexpr BaseScalarType tol = 1.e-10;
+constexpr size_t RNDSEED = 1;
+
+//temp function untill Hermitian containter is implemented
+//** gnerate symmetric-hermitian matrix in a square container */
+template<
+	typename T
+>
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:59]: one line?

File: tests/smoke/alp_zheevd.cpp [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+std::vector< T > generate_symmherm_matrix_data(
+	size_t N,
+	const typename std::enable_if<
+		grb::utils::is_complex< T >::value,
+		void
+	>::type * const = nullptr
+) {
+	std::vector< T > data( N * N );
+	std::fill(data.begin(), data.end(), static_cast< T >( 0 ) );
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:68]: missing space

File: tests/smoke/alp_zheevd.cpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+			data[ j * N + i ] += grb::utils::is_complex< T >::conjugate( data[ i * N + j ] );
+		}
+	}
+	return data;
+}
+
+//** generate upper/lower triangular part of a Symmetric matrix */
+template<
+	typename T
+>
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:83]: one line?

File: tests/smoke/alp_zheevd.cpp [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+constexpr BaseScalarType tol = 1.e-10;
+constexpr size_t RNDSEED = 1;
+
+//temp function untill Hermitian containter is implemented
+//** gnerate symmetric-hermitian matrix in a square container */
+template<
+	typename T
+>
+std::vector< T > generate_symmherm_matrix_data(
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:60]: If these functions are also used in other tests, consider moving them to utils.

File: tests/smoke/alp_zheevd.cpp [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+using HermitianOrSymmetricTridiagonal = structures::SymmetricTridiagonal;
+//fully implemented structures
+using HermitianOrSymmetric = structures::Symmetric;
+#endif
+
+constexpr BaseScalarType tol = 1.e-10;
+constexpr size_t RNDSEED = 1;
+
+//temp function untill Hermitian containter is implemented
+//** gnerate symmetric-hermitian matrix in a square container */
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:56]: /**, capitalize first word.

File: tests/smoke/alp_zheevd.cpp [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+		for( size_t j = i; j < N; ++j ) {
+			//data[ k ] = static_cast< T >( i + j*j ); // easily reproducible
+			data[ k ] = static_cast< T >( std::rand() )  / RAND_MAX;
+			++k;
+		}
+	}
+	return data;
+}
+
+//** check if rows/columns or matrix Q are orthogonal */
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:104]: /**, capitalize first word

File: tests/smoke/alp_zheevd.cpp [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+	const Minus &minus = Minus()
+) {
+	const Scalar< T > zero( ring.template getZero< T >() );
+	const Scalar< T > one( ring.template getOne< T >() );
+
+	RC rc = SUCCESS;
+	const size_t n = nrows( Q );
+
+	// check if QxQt == I
+	alp::Matrix< T, Structure, alp::Density::Dense, ViewType > Qtmp( n );
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:124]: If Stucture is not square, the constructor needs two arguments. Either ensure that the structure is square or add support for non-square structures (whatever makes more sense in this particular case)

File: tests/smoke/alp_zheevd.cpp [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+	rc = rc ? rc : set( Qtmp, zero );
+	rc = rc ? rc : mxm(
+		Qtmp,
+		Q,
+		conjugate( alp::get_view< alp::view::transpose >( Q ) ),
+		ring
+	);
+	Matrix< T, Structure, Dense > Identity( n );
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:132]: Shouldn't the Structure for identity matrix be a predefined one? At least Structure needs to be Symmetric, but this is not ensured.

File: tests/smoke/alp_zheevd.cpp [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+	if( tol < std::abs( fnorm ) ) {
+		std::cout << "The Frobenius norm is too large: " << std::abs( fnorm ) << ".\n";
+		return FAILED;
+	}
+
+	return rc;
+}
+
+
+//** check solution by calculating A x Q - Q x diag(d) */
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:162]: /**, capitalize first letter.

File: tests/smoke/alp_zheevd.cpp [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+		return FAILED;
+	}
+
+	return rc;
+}
+
+
+
+
+void alp_program( const size_t & unit, alp::RC & rc ) {
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:247]: No spaces between & and variable.

File: tests/smoke/alp_zheevd.cpp [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+	// 	std::cout << "Error: mratrix Q is not orthogonal\n";
+	// }
+
+	// rc = check_solution( H, Q, d );
+	// if( rc != SUCCESS ) {
+	// 	std::cout << "Error: solution numerically wrong\n";
+	// }
+}
+
+int main( int argc, char ** argv ) {
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:306]: No spaces between ** and argv.

File: tests/smoke/alp_zheevd.cpp [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+		std::cout << "The Frobenius norm is too large: " << std::abs( fnorm ) << ".\n";
+		return FAILED;
+	}
+
+	return rc;
+}
+
+
+
+
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:246]: To many empty lines

File: tests/smoke/alp_zheevd.cpp [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+		for( size_t j = i; j < N; ++j ) {
+			T val( std::rand(), std::rand() );
+			data[ i * N + j ] = val / std::abs( val );
+			data[ j * N + i ] += grb::utils::is_complex< T >::conjugate( data[ i * N + j ] );
+		}
+	}
+	return data;
+}
+
+//** generate upper/lower triangular part of a Symmetric matrix */
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:80]: /**, capitalize first word

File: tests/smoke/alp_zheevd.cpp [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+	T fnorm = ring.template getZero< T >();
+	rc = rc ? rc : alp::eWiseLambda(
+		[ &fnorm, &ring ]( const size_t i, const size_t j, T &val ) {
+			(void) i;
+			(void) j;
+			internal::foldl( fnorm, val * val, ring.getAdditiveOperator() );
+		},
+		Qtmp
+	);
+	fnorm = std::sqrt( fnorm );
```
Comment (vdimiczrc) [inline tests/smoke/alp_zheevd.cpp:148]: It seems that this is calculated in many tests. Would it make sense to add this function to utils?

File: tests/smoke/alp_zheevd.cpp (hunk 5 above)
Comment (djelovina) [reply tests/smoke/alp_zheevd.cpp:60]: Since many files will be touched, I will leave this for a separate branch.

**Review by vdimiczrc — APPROVED** [review]

**Pull request merged**